### ISSUE TITLE
Switch from SCons to setup.py

### ISSUE THREAD

**Issue by rmjarvis** [GalSim-developers/GalSim#809]
GalSim is currently kind of a pain to install on some systems.  Most python packages can be installed with pip.  It would be nice to switch GalSim to being able to do that, so `pip install galsim` would go install all the appropriate dependencies along with GalSim. 

The main hurdle as far as making this possible is our boost-python dependency, which [can be hard to install](https://github.com/GalSim-developers/GalSim/wiki/Installation%20FAQ#why-is-scons-unable-to-run-a-program-built-with-boostpython) correctly.  I believe that if we want GalSim to be pip-installable, we need to get rid of our boost-python stuff.  This is a big job.

I'd like to replace all the boost-python wrapping with cffi, which I use in TreeCorr to wrap its C++ code.  It is very straightforward if all the connections between Python and C++ are through C functions.  In Python, cffi makes calling the C function extremely easy.  Then in C++ files we have the C functions call out to the regular C++ functions that use full C++ classes to do whatever they need to do.  This isn't currently how GalSim's middle layer is structured, but I think it would be doable to convert to that.

Here are a list of proposals to effect the conversion (going through the list of files in pysrc):
- Angle.cpp: Make Angle and AngleUnit pure python.  We can enforce all the type checking in the python layer and always pass radians down into C++.  If desired, we could keep a separate Angle class in C++ that the C function could convert back into, but I doubt that will generally be required.
- Bessel.cpp: These are already just functions, so that's easy to switch.
- Bounds.cpp: The essence of this can be converted to a C struct, which cffi can handle.  All the Python-layer functionality can be implemented in pure Python, and this struct can pass down into C with an image.
- CDModel.cpp: This is already just a function.  Easy to wrap.
- CorrelatedNoise.cpp: The only thing here is a deprecated function.  We can just remove it, since I expect the pip-installable version can be considered version 2.0, and all 1.x deprecated functions removed.
- HSM.cpp: HSMParams and ShapeData can be turned into C structs.  FindAdaptiveMom and EstimateShear are already just functions, so easy.
- Image.cpp: This is the first difficult one.  I think we can reorg some of the code so that Images are always constructed in Python, and we use a numpy array for the underlying data.  Then these can be passed down into C++ via `double*` pointers and the shape information.  We may want to keep a C++-layer Image class as well, which the middle layer can reconstitute before passing it on to C++ functions that work on it.
- Integ.cpp: We may need to sever the connection between these.  We can keep int1d in C++, but it's probably simplest to just switch any python integrations to a scipy integrator rather than try to handle the call-backs from C++ back to Python functions.  OTOH, I might feel inspired to try to figure it out...
- Interpolant.cpp: I think in the Python layer, we don't really ever do anything with these other than pass them around, so it would be fine to make them opaque `void *` C objects that Python can pass around and the middle layer can reconstitute as needed.
- Noise.cpp: I think all the Noise classes should be made pure python and have various functions like `applyPoissonNoise(rng, im.array)` that could be done in C++.  I don't think we have a strong need for any Noise classes in the C++ layer.
- Random.cpp: This one may require a bit of thought how best to implement.  Probably could have Python classes that hold a `void*` pointer to the C++ version of the class and implement the various methods as functions in which the C function would `reinterpret_cast` back into the actual C++ object.  It's a little tedious, but it's how I handled the Field classes in TreeCorr, so I think I know what to do.  However, there may be some sticky points to figure out.
- SB*.cpp: I think in most cases, we can switch the various GSObjects from holding a SBProfile object and calling methods of it to directly calling C functions.  e.g. Rather than `SBProfile.xValue(x,y)`, Gaussian would have `def xValue(self, x, y): return lib.Gaussian_xValue(sigma,x,y) * flux`.  In some cases, where it's helpful to have a stored object (e.g. InterpolatedImage), there might need to be an opaque C++ object that gets built at initialization and passed down to these functions.
- SBProfile.cpp: The draw, drawK, and drawShoot functions will need some extra work.  I think we'd want to keep `fillXGrid`, `fillKGrid`, and `shoot` implemented in C++ for each of the different kinds of GSObject, but most of the front end work in the draw functions could probably be moved into Python.  These are complicated functions, so I anticipate running into some obstacles with this.  In fact, this bit here is the main reason I haven't gotten around to trying to do all this yet.
- Table.cpp: The code in this file already looks a lot like what we'll need for cffi.  Most things here are already wrapped via an intermediary function.  So these would look similar in the cffi implementation.

**Comment by jmeyers314** [comment]
Is pip synonymous with using setup.py?  I was a bit confused by the title of this issue.

**Comment by rmjarvis** [comment]
Sort of.  pip requires setup.py.  But you can install with setup.py directly without pip.  e.g.
```
python setup.py install --prefix=~
```

**Comment by esheldon** [comment]
++ This would be very helpful to us end users.


On 10/4/16, Mike Jarvis <notifications@github.com> wrote:
> GalSim is currently kind of a pain to install on some systems.  Most python
> packages can be installed with pip.  It would be nice to switch GalSim to
> being able to do that, so `pip install galsim` would go install all the
> appropriate dependencies along with GalSim.
>
> The main hurdle as far as making this possible is our boost-python
> dependency, which [can be hard to
> install](https://github.com/GalSim-developers/GalSim/wiki/Installation%20FAQ#why-is-scons-unable-to-run-a-program-built-with-boostpython)
> correctly.  I believe that if we want GalSim to be pip-installable, we need
> to get rid of our boost-python stuff.  This is a big job.
>
> I'd like to replace all the boost-python wrapping with cffi, which I use in
> TreeCorr to wrap its C++ code.  It is very straightforward if all the
> connections between Python and C++ are through C functions.  In Python, cffi
> makes calling the C function extremely easy.  Then in C++ files we have the
> C functions call out to the regular C++ functions that use full C++ classes
> to do whatever they need to do.  This isn't currently how GalSim's middle
> layer is structured, but I think it would be doable to convert to that.
>
> Here are a list of proposals to effect the conversion (going through the
> list of files in pysrc):
>
> * Angle.cpp: Make Angle and AngleUnit pure python.  We can enforce all the
> type checking in the python layer and always pass radians down into C++.  If
> desired, we could keep a separate Angle class in C++ that the C function
> could convert back into, but I doubt that will generally be required.
> * Bessel.cpp: These are already just functions, so that's easy to switch.
> * Bounds.cpp: The essence of this can be converted to a C struct, which cffi
> can handle.  All the Python-layer functionality can be implemented in pure
> Python, and this struct can pass down into C with an image.
> * CDModel.cpp: This is already just a function.  Easy to wrap.
> * CorrelatedNoise.cpp: The only thing here is a deprecated function.  We can
> just remove it, since I expect the pip-installable version can be considered
> version 2.0, and all 1.x deprecated functions removed.
> * HSM.cpp: HSMParams and ShapeData can be turned into C structs.
> FindAdaptiveMom and EstimateShear are already just functions, so easy.
> * Image.cpp: This is the first difficult one.  I think we can reorg some of
> the code so that Images are always constructed in Python, and we use a numpy
> array for the underlying data.  Then these can be passed down into C++ via
> `double*` pointers and the shape information.  We may want to keep a
> C++-layer Image class as well, which the middle layer can reconstitute
> before passing it on to C++ functions that work on it.
> * Integ.cpp: We may need to sever the connection between these.  We can keep
> int1d in C++, but it's probably simplest to just switch any python
> integrations to a scipy integrator rather than try to handle the call-backs
> from C++ back to Python functions.  OTOH, I might feel inspired to try to
> figure it out...
> * Interpolant.cpp: I think in the Python layer, we don't really ever do
> anything with these other than pass them around, so it would be fine to make
> them opaque `void *` C objects that Python can pass around and the middle
> layer can reconstitute as needed.
> * Noise.cpp: I think all the Noise classes should be made pure python and
> have various functions like `applyPoissonNoise(rng, im.array)` that could be
> done in C++.  I don't think we have a strong need for any Noise classes in
> the C++ layer.
> * Random.cpp: This one may require a bit of thought how best to implement.
> Probably could have Python classes that hold a `void*` pointer to the C++
> version of the class and implement the various methods as functions in which
> the C function would `reinterpret_cast` back into the actual C++ object.
> It's a little tedious, but it's how I handled the Field classes in TreeCorr,
> so I think I know what to do.  However, there may be some sticky points to
> figure out.
> * SB*.cpp: I think in most cases, we can switch the various GSObjects from
> holding a SBProfile object and calling methods of it to directly calling C
> functions.  e.g. Rather than `SBProfile.xValue(x,y)`, Gaussian would have
> `def xValue(self, x, y): return lib.Gaussian_xValue(sigma,x,y) * flux`.  In
> some cases, where it's helpful to have a stored object (e.g.
> InterpolatedImage), there might need to be an opaque C++ object that gets
> built at initialization and passed down to these functions.
> * SBProfile.cpp: The draw, drawK, and drawShoot functions will need some
> extra work.  I think we'd want to keep `fillXGrid`, `fillKGrid`, and `shoot`
> implemented in C++ for each of the different kinds of GSObject, but most of
> the front end work in the draw functions could probably be moved into
> Python.  These are complicated functions, so I anticipate running into some
> obstacles with this.  In fact, this bit here is the main reason I haven't
> gotten around to trying to do all this yet.
> * Table.cpp: The code in this file already looks a lot like what we'll need
> for cffi.  Most things here are already wrapped via an intermediary
> function.  So these would look similar in the cffi implementation.
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/GalSim-developers/GalSim/issues/809


-- 
Erin Scott Sheldon
Brookhaven National Laboratory erin dot sheldon at gmail dot com

**Comment by danielgruen** [comment]
FWIW, the SConscript file is currently broken on macOS Sierra (because it expects the macOS version to be XX.YY.ZZ while Sierra is 10.12).

This is a diff for quick fixing if you stumble across that:

diff --git a/SConstruct b/SConstruct
index 12a529e..aa1ae3e 100644
--- a/SConstruct
+++ b/SConstruct
@@ -225,7 +225,8 @@ def ErrorExit(*args, **kwargs):
     # Give a helpful message if running El Capitan.
     if sys.platform.find('darwin') != -1:
         import platform
-        major, minor, rev = platform.mac_ver()[0].split('.')
+        major = platform.mac_ver()[0].split('.')[0]
+        minor = platform.mac_ver()[0].split('.')[1]
         print 'Mac version: ',major,minor
         if int(major) > 10 or int(minor) >= 11:
             print
@@ -1610,7 +1611,8 @@ int main()
             # Don't require the user to have xcode installed.
             xcode_version = None
             print 'Unable to determine XCode version'
-        major, minor, rev = platform.mac_ver()[0].split('.')
+        major = platform.mac_ver()[0].split('.')[0]
+        minor = platform.mac_ver()[0].split('.')[1]
         if ((int(major) > 10 or int(minor) >= 7) and '-latlas' not in tmv_link and
                 ('-lblas' in tmv_link or '-lcblas' in tmv_link)):
             print 'WARNING: The Apple BLAS library has been found not to be thread safe on'

**Comment by danielgruen** [comment]
Oh, never mind -- this was release/1.3, it's fixed now.

**Comment by rmjarvis** [comment]
Yep.  cf. #750 (in particular the post-closure comments)  Incorporated into v1.4.

**Comment by rmjarvis** [comment]
I've looked into pybind11 a bit more, and apparently it is now pip installable, which means that we could use this within setup.py.  It's based on boost python, so converting over would probably be pretty easy actually.

The caveat is that it requires a c++ compiler that works with c++11.  This means at least g++4.7 or maybe 4.8.  (There are a few c++11 features lacking in 4.7, but I don't know if pybind11 requires them.)

I'm loath to lose the ability to run GalSim on g++4.4, which is still rather common (cf. #849).  But perhaps this is ok.  I've been assuming that the first setup.py release would be 2.0.  So we could keep backporting features to 1.x, for a while at least.  And then for most systems that have g++4.8+, pip install galsim could work.

Thoughts?

**Comment by rmandelb** [comment]
@rmjarvis - a few comments / questions:

1) I was wondering if you could clarify a few things about how you plan to carry out the development on here.  You created a "noboost" branch into which you want to merge this branch (and then carry on with doing more elements of your list in this issue on branch 809).   What is currently on noboost that's not on master?  Will unrelated added functionality that is being added to master in the coming months also be merged into noboost with whatever updates are needed to accommodate the differences between noboost and master?  Or will "noboost" and master be unrelated parallel lines of development for a while, and then there will be one big job to add the new stuff from master into noboost once the noboost work is done?

2) Let me make sure I understand the implications of using pybind11.  Let's say we have someone who is still actively using a g++ version below 4.8.  So this is simply a statement that they'd have to install an updated version of g++ before doing ```pip install galsim``` or trying to directly install GalSim using setup.py?  How hard is that?  My own recollection of when I installed an updated compiler was that it was pretty easy, unlike installing boost, so this change is still potentially a plus for new GalSim users (but I'm not sure how typical my experience was).

**Comment by rmjarvis** [comment]
1. Sure.  So far the only difference between the noboost branch and master is that on the noboost branch I added a line to `.travis.yml` to get travis to do its checks for merges into noboost.  

 My plan was to try to keep noboost up to date with changes in master.  I don't think it will be particularly hard to do, but if anyone develops things that involve the python wrapping stuff, then I'll fix them to make them work in the new scheme as it develops.

 Eventually, the noboost branch would become release 2.0 once everything is finished.  Ideally with the same functionality as the last 1.x version except that the deprecated features will be gone.

2. I don't know enough about pybind11 myself yet to fully know all the implications.  But in my first foray into trying it out, I had SCons detect whether the `-std=c++11` flag works, and if so use pybind11, and otherwise use boost (by adding `USE_BOOST_PYTHON` to CPPDEFINES).

 Then I was able to get some of the classes to work just by setting up `#define`s to normalize the API between the two.  e.g. pybind11 enables pickling by default, so for pybind11 I have `#define enable_pickling` (i.e. replace it with nothing).  They also have different commands for adding properties, so I used a `#define` to alias one of the names to the other one.  Things like that.

 I stopped when I got to pysrc/Image.cpp, since the number of API changes required was kind of enormous.  Mostly in the numpy wrapping stuff.  So I decided to stop there and go back to the stuff that is currently in PR that I had started a little while ago.  I think if we simplify what the wrapping code needs to do, it will become relatively trivial to keep both installation methods working.

 If I can get all that working (and I'm optimistic), then my thought was to have pip and setup.py installations use pybind11.  But if people have a system without a modern c++ compiler, then they can keep using the old SCons-based installation with boost.  Or, as you say, install a newer compiler, which isn't *too* painful.

**Comment by rmandelb** [comment]
Both of those sound good.  Thanks for the explanations.

**Comment by rmjarvis** [comment]
A pre-PR preview for @esheldon in particular, but to all interested parties...

The '#809-pybind11' branch is currently working as a setup.py installation.  Theres a lot of doc editing required still and it is based off of '#809f', so I need to wait for that to be merged before doing a PR.  But if anyone wants to check it out now, feel free.

The one concern, for which I want some input from Erin is that it uses pybind11 for the wrapping, which as we know requires a C++-11 compatible compiler.  Erin, you had expressed reservations about this for linux distros that have gcc 4.4, e.g. the BNL astro cluster.

I've tested this there, and it all works using the anaconda gcc rather than the native gcc.  Here is the procedure to get it going (from the GalSim repo directory):
```
conda create -n galsim
source activate galsim
conda install gcc
conda install pip
pip install -r requirements.txt --user
pip install . --user
```
You could of course use another environment name, like my_root (as most such instructions use).  It doesn't have to be called `galsim`.  Either way, that lets you do `conda install gcc` locally, which then works for installing pybind11 stuff.

Once I've uploaded to pypi, I think the last two lines will just be `pip install galsim --user`.

So...  Does this seem acceptable?  Or should I make a stab at trying out cffi instead?

**Comment by esheldon** [comment]
I'm getting this error
```
(galsim-test) [esheldon@astro0033 GalSim] pip install -r requirements.txt --user
Collecting git+git://github.com/wouterboomsma/eigency.git@ed54a61e8143284e243f0a7ddbc5acb4c8bf58bd (from -r requirements.txt (line 17))
  Cloning git://github.com/wouterboomsma/eigency.git (to ed54a61e8143284e243f0a7ddbc5acb4c8bf58bd) to /data/esheldon/tmp/pip-DVHy8O-build
  Could not find a tag or branch 'ed54a61e8143284e243f0a7ddbc5acb4c8bf58bd', assuming commit.
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/data/esheldon/tmp/pip-DVHy8O-build/setup.py", line 5, in <module>
        from Cython.Build import cythonize
    ImportError: No module named Cython.Build
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /data/esheldon/tmp/pip-DVHy8O-build/
```
I installed cython with `conda install cython` and it did not fix the problem

**Comment by rmjarvis** [comment]
Looks like this is a bug in the eigency setup.py file.  They require cython at build time, and they declare it as such with `setup_requires`, but they import cython at the top, so it is needed before setuptools knows about the requirement.  Maybe I'll give them a PR to fix this.

In the meanwhile, I'm not sure why `conda install cython` didn't work.  After you do that, can you `import cython` in a python session?

It looks like I installed my cython with pip, not conda, so maybe that matters.  You could try `pip install cython` to see if that works better.

**Comment by esheldon** [comment]
yes, installing with pip removed that error, but then I get an import error for numpy

But it is very important to an anaconda user to use the numpy installed via conda, not pip

**Comment by rmjarvis** [comment]
I added a conda_requirements.txt to get everything that can be conda installed done first.  Then use pip for the rest.  So I think the following procedure should work:
```
conda create -y -n galsim
source activate galsim
conda install -y -c conda-forge --file conda_requirements.txt
pip install -r requirements.txt
pip install .
```
At least if worked for me at BNL after deleting all my pip installations and starting with no conda environments.  I might try to find an elegant way to not depend on eigency and use the conda Eigen package instead if possible.  But it seems complicated to get all that working together properly.

**Comment by esheldon** [comment]
it works for me now all the way up until the galsim compilation 
[log.txt](https://github.com/GalSim-developers/GalSim/files/1604714/log.txt)

**Comment by rmjarvis** [comment]
Ah.  Good.  Thanks for testing this out.  That's a file that SCons used to make automatically with the right version information in it.  I forgot to add that to the setup.py installation, but it was still in my repo, since I'd previously done a scons installation.  I'll do that now.

**Comment by rmjarvis** [comment]
Done.  I tested it on a clean repo, and all appears to be working properly now.

**Comment by esheldon** [comment]
Compilation now works!  But I cannot import due to a missing module
```
>>> import galsim
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/gpfs01/astro/workarea/esheldon/anaconda-dirs/miniconda2/envs/galsim-test/lib/python2.7/site-packages/galsim/__init__.py", line 99, in <module>
    from .scene import COSMOSCatalog
  File "/gpfs01/astro/workarea/esheldon/anaconda-dirs/miniconda2/envs/galsim-test/lib/python2.7/site-packages/galsim/scene.py", line 28, in <module>
    from .real import RealGalaxy, RealGalaxyCatalog
  File "/gpfs01/astro/workarea/esheldon/anaconda-dirs/miniconda2/envs/galsim-test/lib/python2.7/site-packages/galsim/real.py", line 36, in <module>
    from .chromatic import ChromaticSum
  File "/gpfs01/astro/workarea/esheldon/anaconda-dirs/miniconda2/envs/galsim-test/lib/python2.7/site-packages/galsim/chromatic.py", line 33, in <module>
    from .bandpass import Bandpass
  File "/gpfs01/astro/workarea/esheldon/anaconda-dirs/miniconda2/envs/galsim-test/lib/python2.7/site-packages/galsim/bandpass.py", line 29, in <module>
    from . import meta_data
ImportError: cannot import name meta_data
```

**Comment by rmjarvis** [comment]
Ack.  Sorry.  You're right.  That's not getting made properly on mine either now.  (It used to be.)  I'll investigate.

**Comment by rmjarvis** [comment]
I think everything is working now.  Both at BNL and on my laptop (with clang), it installs everything properly and all tests run without errors.

Still to do:
[] Make `python setup.py test` run the unit tests
[] Try to get compilation to use multiple processes
[] Update INSTALL.md to describe new installation procedure
[] Check for other doc updates that might now be out of date.

**Comment by esheldon** [comment]
I'm seeing a new failure now, at the galsim installation stage `pip install .`  I started with a clean environment

[log.txt](https://github.com/GalSim-developers/GalSim/files/1611854/log.txt)

**Comment by rmjarvis** [comment]
Sorry, it looks like that was a new feature in distutils version 3.5.  I added a check that I think should make it work for versions prior to this.

Mike

**Comment by esheldon** [comment]
Now it looks like it is for some reason using the system gcc rather than that installed with conda
[log.txt](https://github.com/GalSim-developers/GalSim/files/1612068/log.txt)

**Comment by esheldon** [comment]
sorry, that actuall is the version installed: 4.8.5

anyway, it is giving an error

**Comment by rmjarvis** [comment]
Weird.  It didn't use the `-std=c++11` flag the way it was supposed to.  I wonder if that's also related to the distutils version somehow.  Maybe they changed how some of that works.  

Anyway, I tried adding them a different way -- the same way I did for TreeCorr -- which is slightly more cumbersome, but I think should work.

**Comment by esheldon** [comment]
Works for me now; I am able to import and ran some tests (although there were some separate problems there, will make an issue for them)

**Comment by esheldon** [comment]
Note this failed for a student of ours who tried to do it on a mac.  Unfortunately I don't have a full report to give, but it might be good to try it on your mac now if you haven't yet.  It was a new mac so probably latest OS version.

**Comment by rmjarvis** [comment]
I appreciate you trying it out for me, but let's not be spreading this too far and wide just yet.  I'm still developing things here, so it may break on different systems or configurations as I continue to make changes.  Once I submit it as a pull request, then it would be worth asking lots of people to try it out on as many different systems as possible.

**Comment by esheldon** [comment]
understood

**Comment by rmjarvis** [comment]
Finally done.  Via PRs #853, #857, #863, #902, #911, #931, #950.
